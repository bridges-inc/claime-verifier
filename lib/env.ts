const projectName = 'claime-verifier'

export enum Environments {
  PROD = 'prod',
  DEV = 'dev',
  TEST = 'test',
}
export interface EnvironmentVariables {
  allowedOrigin: string
  rootDomain: string
  hostedZoneId: string
  certificateArn: string
}

const EnvironmentVariablesSetting: {
  [key in Environments]: EnvironmentVariables
} = {
  [Environments.PROD]: {
    allowedOrigin: '',
    rootDomain: '',
    hostedZoneId: '',
    certificateArn: '',
  },
  [Environments.DEV]: {
    allowedOrigin: 'https://claime-webfront-k6p1srx99-squard.vercel.app',
    rootDomain: 'claime-dev.tk',
    hostedZoneId: 'Z08620181ARYV5PENJUEI',
    certificateArn:
      'arn:aws:acm:us-east-1:495476032358:certificate/92761d01-6e38-4ca4-8580-a945a5c379fe',
  },
  [Environments.TEST]: {
    allowedOrigin: '',
    rootDomain: 'test',
    hostedZoneId: 'test',
    certificateArn: '',
  },
}

export function valueOf(env: Environments): EnvironmentVariables {
  return EnvironmentVariablesSetting[env]
}
export const withEnvPrefix = (target: Environments, str: string) =>
  `${projectName}-${str}-${target}`

export const isProd = (target: Environments) => {
  return target === Environments.PROD
}
export const isDev = (target: Environments) => {
  return target === Environments.DEV
}
