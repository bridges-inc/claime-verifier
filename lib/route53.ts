import {
  ARecord,
  ARecordProps,
  CnameRecord,
  HostedZone,
  IHostedZone,
  PublicHostedZone,
  RecordTarget,
} from '@aws-cdk/aws-route53'
import { Construct, Stack, StackProps } from '@aws-cdk/core'
import * as environment from './env'

export class Route53Stack extends Stack {
  public readonly hostedZone: IHostedZone
  constructor(
    scope: Construct,
    id: string,
    target: environment.Environments,
    props?: StackProps,
  ) {
    super(scope, id, props)
    const { rootDomain } = environment.valueOf(target)
    if (environment.isProd(target)) {
      return
    }
    this.hostedZone = hostedZone(this, target)
    // Vercel Domain Verification
    const vercelARecordProps: ARecordProps = {
      zone: this.hostedZone,
      recordName: `${rootDomain}`,
      target: RecordTarget.fromIpAddresses('76.76.21.21'),
    }
    new ARecord(this, 'VercelARecord', vercelARecordProps)

    new CnameRecord(this, 'VercelCNAMERecord', {
      zone: this.hostedZone,
      recordName: `www.${rootDomain}`,
      domainName: 'cname.vercel-dns.com',
    })
  }
}

const hostedZone = (
  scope: Construct,
  target: environment.Environments,
): IHostedZone => {
  return new PublicHostedZone(scope, 'HostedZone', {
    zoneName: hostedZoneName(target),
    comment: 'created by cdk',
  })
}

export const hostedZoneName = (target: environment.Environments) => {
  return environment.valueOf(target).rootDomain
}

export const hostedZoneFromId = (
  scope: Construct,
  target: environment.Environments,
) => {
  return HostedZone.fromHostedZoneAttributes(scope, `HostedZone`, {
    zoneName: hostedZoneName(target),
    hostedZoneId: environment.valueOf(target).hostedZoneId,
  })
}
