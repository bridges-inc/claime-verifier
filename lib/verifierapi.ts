import { LambdaIntegration, RestApi } from '@aws-cdk/aws-apigateway'
import { Function, Runtime, Tracing } from '@aws-cdk/aws-lambda'
import * as cdk from '@aws-cdk/core'
import {
  addCorsOptions,
  code,
  environmentVariables,
  withCustomDomain,
} from './api'
import { basicPolicytStatements } from './discord'
import * as environment from './env'

export class VerifierApiStack extends cdk.Stack {
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments,
    props?: cdk.StackProps,
  ) {
    super(scope, id, props)
    const api = new RestApi(this, 'RestApi', {
      restApiName: environment.withEnvPrefix(target, 'verifier'),
    })
    apifunction(this, 'verify', '/verify/{eoa}', target, api)
    apifunction(this, 'testVerification', '/test/verify/{eoa}', target, api)
    withCustomDomain(this, api, restApiDomainName(target), target)
  }
}

const apifunction = (
  stack: cdk.Stack,
  resource: string,
  path: string,
  target: environment.Environments,
  api: RestApi,
) => {
  const func = new Function(stack, resource, {
    functionName: `${environment.withEnvPrefix(target, resource)}`,
    code: code(resource),
    handler: 'bin/main',
    timeout: cdk.Duration.minutes(1),
    runtime: Runtime.GO_1_X,
    environment: environmentVariables(target),
    tracing: Tracing.ACTIVE,
  })
  basicPolicytStatements(stack.region, stack.account, target).forEach((s) =>
    func.addToRolePolicy(s),
  )
  const re = api.root.addResource(path)
  re.addMethod('GET', new LambdaIntegration(func))
  addCorsOptions(re, target)
  return func
}

const restApiDomainName = (target: environment.Environments) => {
  return `verifier.` + environment.valueOf(target).rootDomain
}
