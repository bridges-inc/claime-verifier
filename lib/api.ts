import {
  DomainName,
  EndpointType,
  IResource,
  MockIntegration,
  PassthroughBehavior,
  RestApi,
  SecurityPolicy,
} from '@aws-cdk/aws-apigateway'
import * as certificatemanager from '@aws-cdk/aws-certificatemanager'
import { Code } from '@aws-cdk/aws-lambda'
import * as route53 from '@aws-cdk/aws-route53'
import * as alias from '@aws-cdk/aws-route53-targets'
import * as cdk from '@aws-cdk/core'
import { Stack } from '@aws-cdk/core'
import { resolve } from 'path'
import * as environment from './env'
import { hostedZoneFromId } from './route53'

type CordOptionsConfig = {
  methods: ('GET' | 'POST' | 'PUT' | 'DELETE')[]
  withCredentials?: true
  headers?: string[]
  origin?: string
}
export function addCorsOptions(
  apiResource: IResource,
  config: CordOptionsConfig,
) {
  const responseParameters = integrationResponseParameters(config)
  const methodResponseParameters = Object.keys(responseParameters).reduce(
    (res, key) => Object.assign(res, { [key]: true }),
    {},
  )
  apiResource.addMethod(
    'OPTIONS',
    new MockIntegration({
      integrationResponses: [{ statusCode: '200', responseParameters }],
      passthroughBehavior: PassthroughBehavior.NEVER,
      requestTemplates: { 'application/json': '{"statusCode": 200}' },
    }),
    {
      methodResponses: [
        { statusCode: '200', responseParameters: methodResponseParameters },
      ],
    },
  )
}

const integrationResponseParameters = (config: CordOptionsConfig) => {
  const { methods, withCredentials, headers, origin } = config
  const params = {
    'method.response.header.Access-Control-Allow-Headers': `'${
      headers?.join(',') || '*'
    }'`,
    'method.response.header.Access-Control-Allow-Origin': `'${origin || '*'}'`,
    'method.response.header.Access-Control-Allow-Methods': `'OPTIONS,${methods.join(
      ',',
    )}'`,
  }
  if (withCredentials) {
    Object.assign(params, {
      'method.response.header.Access-Control-Allow-Credentials': "'true'",
    })
  }
  return params
}

export const environmentVariables = (target: environment.Environments) => {
  return {
    AllowedOrigin: environment.valueOf(target).allowedOrigin,
    EnvironmentId: target,
    SubgraphEndpoint: environment.valueOf(target).subgraphEndpoint,
  }
}
const aRecord = (
  stack: Stack,
  target: environment.Environments,
  domainName: string,
  customDomain: DomainName,
) => {
  new route53.ARecord(stack, 'RestApiARecord', {
    zone: hostedZoneFromId(stack, target),
    recordName: domainName,
    target: route53.RecordTarget.fromAlias(
      new alias.ApiGatewayDomain(customDomain),
    ),
  })
}

export const code = (dirname: string) => {
  return Code.fromAsset(
    resolve(`${__dirname}/../`, 'lib', 'functions', dirname, 'bin', 'main.zip'),
  )
}
export const withCustomDomain = (
  stack: cdk.Stack,
  api: RestApi,
  domain: string,
  target: environment.Environments,
) => {
  const customDomain = api.addDomainName(
    environment.withEnvPrefix(target, 'domain'),
    customDomainProps(stack, domain, target),
  )
  aRecord(stack, target, domain, customDomain)
  return customDomain
}

const customDomainProps = (
  stack: cdk.Stack,
  domain: string,
  target: environment.Environments,
) => {
  return {
    domainName: domain,
    certificate: certificatemanager.Certificate.fromCertificateArn(
      stack,
      'Cert',
      environment.valueOf(target).certificateArn,
    ),
    securityPolicy: SecurityPolicy.TLS_1_2,
    endpointType: EndpointType.REGIONAL,
  }
}
