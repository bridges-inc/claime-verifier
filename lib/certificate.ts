import {
  DnsValidatedCertificate,
  ICertificate,
  ValidationMethod,
} from '@aws-cdk/aws-certificatemanager'
import { Construct, Stack } from '@aws-cdk/core'
import * as environment from './env'
import { hostedZoneFromId } from './route53'

export class CertificateStack extends Stack {
  public readonly certificate: ICertificate

  constructor(scope: Construct, id: string, target: environment.Environments) {
    super(scope, id)
    const { hostedZoneId } = environment.valueOf(target)
    if (environment.isProd(target)) return
    if (!hostedZoneId) throw new Error('env.hostedZoneId is requied')
    this.certificate = certificate(this, target)
  }
}

const certificate = (
  scope: Construct,
  target: environment.Environments,
): ICertificate => {
  const { rootDomain } = environment.valueOf(target)

  return new DnsValidatedCertificate(scope, 'Certificate', {
    domainName: `${rootDomain}`,
    subjectAlternativeNames: [`*.${rootDomain}`],
    hostedZone: hostedZoneFromId(scope, target),
    validationMethod: ValidationMethod.DNS,
    region: 'us-east-1',
  })
}
