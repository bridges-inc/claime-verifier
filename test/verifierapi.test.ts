import { SynthUtils } from '@aws-cdk/assert'
import { App } from '@aws-cdk/core'
import { Environments } from '../lib/env'
import { Route53Stack } from '../lib/route53'
import { VerifierApiStack } from '../lib/verifierapi'

test('resource created', () => {
  const app = new App()
  const target = Environments.TEST
  const r53stack = new Route53Stack(app, 'test-hostedZone', target)
  const stack = new VerifierApiStack(app, 'test', target, {
    hostedZone: r53stack.hostedZone,
  })
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
})
