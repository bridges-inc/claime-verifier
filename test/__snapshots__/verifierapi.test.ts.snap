// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resource created 1`] = `
Object {
  "Outputs": Object {
    "RestApiEndpoint0551178A": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "RestApi0C43BF4B",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "[HASH REMOVED]",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
  },
  "Resources": Object {
    "RestApi0C43BF4B": Object {
      "Properties": Object {
        "Name": "claime-verifier-verifier-test",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "RestApiARecord0F033820": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "RestApiclaimeverifierdomaintest8DF44C4D",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::GetAtt": Array [
              "RestApiclaimeverifierdomaintest8DF44C4D",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "",
        "Name": "verifier.test.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "RestApiAccount7C83CF5A": Object {
      "DependsOn": Array [
        "RestApi0C43BF4B",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "RestApiCloudWatchRoleE3ED6605",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "RestApiCloudWatchRoleE3ED6605": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "[HASH REMOVED]": Object {
      "DependsOn": Array [
        "RestApitest9059D171",
        "RestApitestverifyeoaGETA37B8E1D",
        "RestApitestverifyeoaOPTIONSAA88D55F",
        "RestApitestverifyeoa7B79A4AB",
        "RestApitestverify811AADE0",
        "RestApiverifyeoaGET8F9A4083",
        "RestApiverifyeoaOPTIONS1995E63A",
        "RestApiverifyeoa0F5B975F",
        "RestApiverifyBAACE9AF",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "[HASH REMOVED]": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "[HASH REMOVED]",
        },
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "RestApiclaimeverifierdomaintest8DF44C4D": Object {
      "Properties": Object {
        "DomainName": "verifier.test",
        "EndpointConfiguration": Object {
          "Types": Array [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": "",
        "SecurityPolicy": "TLS_1_2",
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "RestApiclaimeverifierdomaintestMaptestRestApi41EB688D8756FA87": Object {
      "Properties": Object {
        "DomainName": Object {
          "Ref": "RestApiclaimeverifierdomaintest8DF44C4D",
        },
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
        "Stage": Object {
          "Ref": "[HASH REMOVED]",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "RestApitest9059D171": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "RestApi0C43BF4B",
            "RootResourceId",
          ],
        },
        "PathPart": "test",
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApitestverify811AADE0": Object {
      "Properties": Object {
        "ParentId": Object {
          "Ref": "RestApitest9059D171",
        },
        "PathPart": "verify",
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApitestverifyeoa7B79A4AB": Object {
      "Properties": Object {
        "ParentId": Object {
          "Ref": "RestApitestverify811AADE0",
        },
        "PathPart": "{eoa}",
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApitestverifyeoaGETA37B8E1D": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "testVerification888035B9",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "RestApitestverifyeoa7B79A4AB",
        },
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApitestverifyeoaGETApiPermissionTesttestRestApi41EB688DGETtestverifyeoaCE74D562": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "testVerification888035B9",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "RestApi0C43BF4B",
              },
              "/test-invoke-stage/GET/test/verify/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApitestverifyeoaGETApiPermissiontestRestApi41EB688DGETtestverifyeoa8813905A": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "testVerification888035B9",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "RestApi0C43BF4B",
              },
              "/",
              Object {
                "Ref": "[HASH REMOVED]",
              },
              "/GET/test/verify/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApitestverifyeoaOPTIONSAA88D55F": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": Object {
          "IntegrationResponses": Array [
            Object {
              "ResponseParameters": Object {
                "method.response.header.Access-Control-Allow-Credentials": "'false'",
                "method.response.header.Access-Control-Allow-Headers": "'*'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,POST'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "200",
            },
          ],
          "PassthroughBehavior": "NEVER",
          "RequestTemplates": Object {
            "application/json": "{"statusCode": 200}",
          },
          "Type": "MOCK",
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Access-Control-Allow-Credentials": true,
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "200",
          },
        ],
        "ResourceId": Object {
          "Ref": "RestApitestverifyeoa7B79A4AB",
        },
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApiverifyBAACE9AF": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "RestApi0C43BF4B",
            "RootResourceId",
          ],
        },
        "PathPart": "verify",
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApiverifyeoa0F5B975F": Object {
      "Properties": Object {
        "ParentId": Object {
          "Ref": "RestApiverifyBAACE9AF",
        },
        "PathPart": "{eoa}",
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApiverifyeoaGET8F9A4083": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "verify2EA97391",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "RestApiverifyeoa0F5B975F",
        },
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApiverifyeoaGETApiPermissionTesttestRestApi41EB688DGETverifyeoaAA0DA3EA": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "verify2EA97391",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "RestApi0C43BF4B",
              },
              "/test-invoke-stage/GET/verify/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiverifyeoaGETApiPermissiontestRestApi41EB688DGETverifyeoaABDB1EB7": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "verify2EA97391",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "RestApi0C43BF4B",
              },
              "/",
              Object {
                "Ref": "[HASH REMOVED]",
              },
              "/GET/verify/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiverifyeoaOPTIONS1995E63A": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": Object {
          "IntegrationResponses": Array [
            Object {
              "ResponseParameters": Object {
                "method.response.header.Access-Control-Allow-Credentials": "'false'",
                "method.response.header.Access-Control-Allow-Headers": "'*'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,POST'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "200",
            },
          ],
          "PassthroughBehavior": "NEVER",
          "RequestTemplates": Object {
            "application/json": "{"statusCode": 200}",
          },
          "Type": "MOCK",
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Access-Control-Allow-Credentials": true,
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "200",
          },
        ],
        "ResourceId": Object {
          "Ref": "RestApiverifyeoa0F5B975F",
        },
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "testVerification888035B9": Object {
      "DependsOn": Array [
        "testVerificationServiceRoleDefaultPolicyC53FEB41",
        "testVerificationServiceRole04D8062F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "[HASH REMOVED]",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AllowedOrigin": "",
            "EnvironmentId": "test",
            "SubgraphEndpoint": "endpoint-of-subgraph",
          },
        },
        "FunctionName": "claime-verifier-testVerification-test",
        "Handler": "bin/main",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testVerificationServiceRole04D8062F",
            "Arn",
          ],
        },
        "Runtime": "go1.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "testVerificationServiceRole04D8062F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testVerificationServiceRoleDefaultPolicyC53FEB41": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:Put*",
                "dynamodb:Get*",
                "dynamodb:Scan*",
                "dynamodb:Query*",
                "dynamodb:Delete*",
                "dynamodb:Batch*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/claime-verifier-main-test*",
                  ],
                ],
              },
            },
            Object {
              "Action": "ssm:Get*",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/claime-verifier*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testVerificationServiceRoleDefaultPolicyC53FEB41",
        "Roles": Array [
          Object {
            "Ref": "testVerificationServiceRole04D8062F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "verify2EA97391": Object {
      "DependsOn": Array [
        "verifyServiceRoleDefaultPolicyBE827A6A",
        "verifyServiceRole4DDB2A47",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "[HASH REMOVED]",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AllowedOrigin": "",
            "EnvironmentId": "test",
            "SubgraphEndpoint": "endpoint-of-subgraph",
          },
        },
        "FunctionName": "claime-verifier-verify-test",
        "Handler": "bin/main",
        "Role": Object {
          "Fn::GetAtt": Array [
            "verifyServiceRole4DDB2A47",
            "Arn",
          ],
        },
        "Runtime": "go1.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "verifyServiceRole4DDB2A47": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "verifyServiceRoleDefaultPolicyBE827A6A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:Put*",
                "dynamodb:Get*",
                "dynamodb:Scan*",
                "dynamodb:Query*",
                "dynamodb:Delete*",
                "dynamodb:Batch*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/claime-verifier-main-test*",
                  ],
                ],
              },
            },
            Object {
              "Action": "ssm:Get*",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/claime-verifier*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "verifyServiceRoleDefaultPolicyBE827A6A",
        "Roles": Array [
          Object {
            "Ref": "verifyServiceRole4DDB2A47",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
